invariance.md

```diff
@@ -0,0 +1,204 @@
+>
+> #### Reason
+>
+> Readability.
+> Ease of comprehension.
+> The use of `class` alerts the programmer to the need for an invariant.
+> This is a useful convention.
+>
+> #### Note
+>
+> An invariant is a logical condition for the members of an object that a constructor must establish for the public member functions to assume.
+> After the invariant is established (typically by a constructor) every member function can be called for the object.
+> An invariant can be stated informally (e.g., in a comment) or more formally using `Expects`.
+>
+> If all data members can vary independently of each other, no invariant is possible.
+>
+> #### Example
+>
+> ```cpp
+>     struct Pair {  // the members can vary independently
+>         string name;
+>         int volume;
+>     };
+> ```
+>
+> but:
+>
+> ```cpp
+>     class Date {
+>     public:
+>         // validate that {yy, mm, dd} is a valid date and initialize
+>         Date(int yy, Month mm, char dd);
+>         // ...
+>     private:
+>         int y;
+>         Month m;
+>         char d;    // day
+>     };
+> ```
+>
+> #### Note
+>
+> If a class has any `private` data, a user cannot completely initialize an object without the use of a constructor.
+> Hence, the class definer will provide a constructor and must specify its meaning.
+> This effectively means the definer needs to define an invariant.
+>
+> **See also**:
+>
+> * [define a class with private data as `class`](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-class)
+> * [Prefer to place the interface first in a class](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rl-order)
+> * [minimize exposure of members](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-private)
+> * [Avoid `protected` data](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-protected)
+>
+> #### Enforcement
+>
+> Look for `struct`s with all data private and `class`es with public members.
+
+## But what about when my invariants or data change?
+
+One of the concerns that can arise from choosing to put data into a struct instead of a class with getters and setters is that those methods provide an interface that insulates callers if the underlying data needs to change. I agree that this not only can but absolutely *will* happen at some point. However, I think a couple of things can serve as protections. First, [DMI](dmi.md) gives guidance that keeps the scope of data small, meaning that the scope of changes will also be small. Second, changing from setting a variable directly to calling a getter or setter is a minor refactor. Third, the placement of data is often just a best guess at first, so the location, type, etc. of data is likely to change over time anyway. The setter/getter interface is also fluid in the first implementation. Modern IDEs continue to make this easier, and the introduction of AI will only continue this trend.
```

Comment:
This first sentence is confusing to me. It sounds like "one of the concerns [...] is that those methods provide an interface that insulates callers if the underlying data needs to change" but the choice was putting "data into a struct instead of a class". So it reads as the problem is the struct because the class has methods that insulate callers from change, and I don't get that. From there I don't know what point the rest of the paragraph is making until we get to your "second" point, which suggests you're advocating converting structs to classes.
I don't know that this answers the question in the heading, and I don't have in my head a thought of a situation where that might occur off hand.

Reply:
I can completely see where I wasn't clear in my wording. I rewrote this as a more explicit problem paragraph and left most of the solution the same.